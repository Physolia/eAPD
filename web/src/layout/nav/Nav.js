import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { VerticalNav } from '@cmsgov/design-system';
import { connect } from 'react-redux';
import NavLink from './NavLink';
import { generateKey as actualGenerateKey } from '../../util';

const Nav = ({ generateKey, items, pathname }) => {
  // force component update when pathname changes
  const [key, setKey] = useState('');
  useEffect(() => setKey(generateKey()), [pathname, generateKey]);
  return (
    <nav aria-label="Main Navigation">
      <VerticalNav component={NavLink} items={items} key={key} />
    </nav>
  );
};

Nav.defaultProps = {
  generateKey: actualGenerateKey
};

Nav.propTypes = {
  generateKey: PropTypes.func,
  items: PropTypes.array.isRequired,
  pathname: PropTypes.string.isRequired
};

const mapStateToProps = ({ nav, router }) => ({
  items: nav.items,
  pathname: router.location.pathname
});

export default connect(mapStateToProps)(Nav);

export { Nav as plain, NavLink, mapStateToProps };
