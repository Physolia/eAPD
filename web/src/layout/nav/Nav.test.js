import { mount } from 'enzyme';
import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';

import { plain as Nav, mapStateToProps } from './Nav';

describe('<Nav /> component', () => {
  it('renders correctly', () => {
    const props = {
      generateKey: () => 'asdf1234',
      items: [
        { id: 'home-nav', label: 'Home', url: '/' },
        { id: 'apd-nav', label: 'View APD', url: '/apd', selected: true },
        { id: 'search-nav', label: 'Search', url: '/search' },
        {
          id: 'settings-nav',
          label: 'Settings',
          items: [
            { id: 'profile-nav', label: 'Profile', url: '/me' },
            { id: 'logout-nav', label: 'Log out', url: '/logout' }
          ]
        }
      ],
      pathname: '/apd'
    };
    const component = mount(
      <Router>
        <Nav {...props} />
      </Router>
    );
    expect(component).toMatchSnapshot();
  });

  it('maps state to props', () => {
    const state = {
      nav: {
        items: ['home-nav', 'search-nav', 'settings-nav']
      },
      router: {
        location: {
          pathname: '/apd'
        }
      }
    };

    expect(mapStateToProps(state)).toEqual({
      items: ['home-nav', 'search-nav', 'settings-nav'],
      pathname: '/apd'
    });
  });
});
