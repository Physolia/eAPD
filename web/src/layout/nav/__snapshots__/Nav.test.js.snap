// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav /> component renders correctly 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Nav
      generateKey={[Function]}
      items={
        Array [
          Object {
            "id": "home-nav",
            "label": "Home",
            "url": "/",
          },
          Object {
            "id": "apd-nav",
            "label": "View APD",
            "selected": true,
            "url": "/apd",
          },
          Object {
            "id": "search-nav",
            "label": "Search",
            "url": "/search",
          },
          Object {
            "id": "settings-nav",
            "items": Array [
              Object {
                "id": "profile-nav",
                "label": "Profile",
                "url": "/me",
              },
              Object {
                "id": "logout-nav",
                "label": "Log out",
                "url": "/logout",
              },
            ],
            "label": "Settings",
          },
        ]
      }
      pathname="/apd"
    >
      <nav
        aria-label="Main Navigation"
      >
        <VerticalNav
          collapsed={false}
          component={[Function]}
          items={
            Array [
              Object {
                "id": "home-nav",
                "label": "Home",
                "url": "/",
              },
              Object {
                "id": "apd-nav",
                "label": "View APD",
                "selected": true,
                "url": "/apd",
              },
              Object {
                "id": "search-nav",
                "label": "Search",
                "url": "/search",
              },
              Object {
                "id": "settings-nav",
                "items": Array [
                  Object {
                    "id": "profile-nav",
                    "label": "Profile",
                    "url": "/me",
                  },
                  Object {
                    "id": "logout-nav",
                    "label": "Log out",
                    "url": "/logout",
                  },
                ],
                "label": "Settings",
              },
            ]
          }
          key="asdf1234"
        >
          <nav>
            <ul
              className="ds-c-vertical-nav"
            >
              <VerticalNavItem
                ariaCollapsedStateButtonLabel="Expand sub-navigation"
                ariaExpandedStateButtonLabel="Collapse sub-navigation"
                component={[Function]}
                defaultCollapsed={false}
                id="home-nav"
                key="home-nav/Home"
                label="Home"
                url="/"
              >
                <li
                  className="ds-c-vertical-nav__item"
                >
                  <VerticalNavItemLabel
                    ariaCollapsedStateButtonLabel="Expand sub-navigation"
                    ariaExpandedStateButtonLabel="Collapse sub-navigation"
                    collapsed={false}
                    component={[Function]}
                    hasSubnav={false}
                    label="Home"
                    onClick={[Function]}
                    selected={false}
                    subnavId="home-nav__subnav"
                    url="/"
                  >
                    <NavLink
                      className="ds-c-vertical-nav__label"
                      href="/"
                      onClick={[Function]}
                    >
                      <Link
                        className="ds-c-vertical-nav__label"
                        onClick={[Function]}
                        to="/"
                      >
                        <LinkAnchor
                          className="ds-c-vertical-nav__label"
                          href="/"
                          navigate={[Function]}
                          onClick={[Function]}
                        >
                          <a
                            className="ds-c-vertical-nav__label"
                            href="/"
                            onClick={[Function]}
                          >
                            Home
                          </a>
                        </LinkAnchor>
                      </Link>
                    </NavLink>
                  </VerticalNavItemLabel>
                </li>
              </VerticalNavItem>
              <VerticalNavItem
                ariaCollapsedStateButtonLabel="Expand sub-navigation"
                ariaExpandedStateButtonLabel="Collapse sub-navigation"
                component={[Function]}
                defaultCollapsed={false}
                id="apd-nav"
                key="apd-nav/apdView APD"
                label="View APD"
                selected={true}
                url="/apd"
              >
                <li
                  className="ds-c-vertical-nav__item"
                >
                  <VerticalNavItemLabel
                    ariaCollapsedStateButtonLabel="Expand sub-navigation"
                    ariaExpandedStateButtonLabel="Collapse sub-navigation"
                    collapsed={false}
                    component={[Function]}
                    hasSubnav={false}
                    label="View APD"
                    onClick={[Function]}
                    selected={true}
                    subnavId="apd-nav__subnav"
                    url="/apd"
                  >
                    <NavLink
                      className="ds-c-vertical-nav__label ds-c-vertical-nav__label--current"
                      href="/apd"
                      onClick={[Function]}
                    >
                      <Link
                        className="ds-c-vertical-nav__label ds-c-vertical-nav__label--current"
                        onClick={[Function]}
                        to="/apd"
                      >
                        <LinkAnchor
                          className="ds-c-vertical-nav__label ds-c-vertical-nav__label--current"
                          href="/apd"
                          navigate={[Function]}
                          onClick={[Function]}
                        >
                          <a
                            className="ds-c-vertical-nav__label ds-c-vertical-nav__label--current"
                            href="/apd"
                            onClick={[Function]}
                          >
                            View APD
                          </a>
                        </LinkAnchor>
                      </Link>
                    </NavLink>
                  </VerticalNavItemLabel>
                </li>
              </VerticalNavItem>
              <VerticalNavItem
                ariaCollapsedStateButtonLabel="Expand sub-navigation"
                ariaExpandedStateButtonLabel="Collapse sub-navigation"
                component={[Function]}
                defaultCollapsed={false}
                id="search-nav"
                key="search-nav/searchSearch"
                label="Search"
                url="/search"
              >
                <li
                  className="ds-c-vertical-nav__item"
                >
                  <VerticalNavItemLabel
                    ariaCollapsedStateButtonLabel="Expand sub-navigation"
                    ariaExpandedStateButtonLabel="Collapse sub-navigation"
                    collapsed={false}
                    component={[Function]}
                    hasSubnav={false}
                    label="Search"
                    onClick={[Function]}
                    selected={false}
                    subnavId="search-nav__subnav"
                    url="/search"
                  >
                    <NavLink
                      className="ds-c-vertical-nav__label"
                      href="/search"
                      onClick={[Function]}
                    >
                      <Link
                        className="ds-c-vertical-nav__label"
                        onClick={[Function]}
                        to="/search"
                      >
                        <LinkAnchor
                          className="ds-c-vertical-nav__label"
                          href="/search"
                          navigate={[Function]}
                          onClick={[Function]}
                        >
                          <a
                            className="ds-c-vertical-nav__label"
                            href="/search"
                            onClick={[Function]}
                          >
                            Search
                          </a>
                        </LinkAnchor>
                      </Link>
                    </NavLink>
                  </VerticalNavItemLabel>
                </li>
              </VerticalNavItem>
              <VerticalNavItem
                ariaCollapsedStateButtonLabel="Expand sub-navigation"
                ariaExpandedStateButtonLabel="Collapse sub-navigation"
                component={[Function]}
                defaultCollapsed={false}
                id="settings-nav"
                items={
                  Array [
                    Object {
                      "id": "profile-nav",
                      "label": "Profile",
                      "url": "/me",
                    },
                    Object {
                      "id": "logout-nav",
                      "label": "Log out",
                      "url": "/logout",
                    },
                  ]
                }
                key="settings-navundefinedSettings"
                label="Settings"
              >
                <li
                  className="ds-c-vertical-nav__item"
                >
                  <VerticalNavItemLabel
                    ariaCollapsedStateButtonLabel="Expand sub-navigation"
                    ariaExpandedStateButtonLabel="Collapse sub-navigation"
                    collapsed={false}
                    component={[Function]}
                    hasSubnav={true}
                    label="Settings"
                    onClick={[Function]}
                    selected={false}
                    subnavId="settings-nav__subnav"
                  >
                    <button
                      aria-controls="settings-nav__subnav"
                      aria-expanded={true}
                      className="ds-c-vertical-nav__label ds-c-vertical-nav__label--parent"
                      onClick={[Function]}
                      title="Collapse sub-navigation"
                    >
                      Settings
                    </button>
                  </VerticalNavItemLabel>
                  <VerticalNav
                    collapsed={false}
                    component={[Function]}
                    id="settings-nav__subnav"
                    items={
                      Array [
                        Object {
                          "id": "profile-nav",
                          "label": "Profile",
                          "url": "/me",
                        },
                        Object {
                          "id": "logout-nav",
                          "label": "Log out",
                          "url": "/logout",
                        },
                      ]
                    }
                    nested={true}
                  >
                    <nav>
                      <ul
                        className="ds-c-vertical-nav__subnav"
                        id="settings-nav__subnav"
                      >
                        <VerticalNavItem
                          ariaCollapsedStateButtonLabel="Expand sub-navigation"
                          ariaExpandedStateButtonLabel="Collapse sub-navigation"
                          component={[Function]}
                          defaultCollapsed={false}
                          id="profile-nav"
                          key="profile-nav/meProfile"
                          label="Profile"
                          url="/me"
                        >
                          <li
                            className="ds-c-vertical-nav__item"
                          >
                            <VerticalNavItemLabel
                              ariaCollapsedStateButtonLabel="Expand sub-navigation"
                              ariaExpandedStateButtonLabel="Collapse sub-navigation"
                              collapsed={false}
                              component={[Function]}
                              hasSubnav={false}
                              label="Profile"
                              onClick={[Function]}
                              selected={false}
                              subnavId="profile-nav__subnav"
                              url="/me"
                            >
                              <NavLink
                                className="ds-c-vertical-nav__label"
                                href="/me"
                                onClick={[Function]}
                              >
                                <Link
                                  className="ds-c-vertical-nav__label"
                                  onClick={[Function]}
                                  to="/me"
                                >
                                  <LinkAnchor
                                    className="ds-c-vertical-nav__label"
                                    href="/me"
                                    navigate={[Function]}
                                    onClick={[Function]}
                                  >
                                    <a
                                      className="ds-c-vertical-nav__label"
                                      href="/me"
                                      onClick={[Function]}
                                    >
                                      Profile
                                    </a>
                                  </LinkAnchor>
                                </Link>
                              </NavLink>
                            </VerticalNavItemLabel>
                          </li>
                        </VerticalNavItem>
                        <VerticalNavItem
                          ariaCollapsedStateButtonLabel="Expand sub-navigation"
                          ariaExpandedStateButtonLabel="Collapse sub-navigation"
                          component={[Function]}
                          defaultCollapsed={false}
                          id="logout-nav"
                          key="logout-nav/logoutLog out"
                          label="Log out"
                          url="/logout"
                        >
                          <li
                            className="ds-c-vertical-nav__item"
                          >
                            <VerticalNavItemLabel
                              ariaCollapsedStateButtonLabel="Expand sub-navigation"
                              ariaExpandedStateButtonLabel="Collapse sub-navigation"
                              collapsed={false}
                              component={[Function]}
                              hasSubnav={false}
                              label="Log out"
                              onClick={[Function]}
                              selected={false}
                              subnavId="logout-nav__subnav"
                              url="/logout"
                            >
                              <NavLink
                                className="ds-c-vertical-nav__label"
                                href="/logout"
                                onClick={[Function]}
                              >
                                <Link
                                  className="ds-c-vertical-nav__label"
                                  onClick={[Function]}
                                  to="/logout"
                                >
                                  <LinkAnchor
                                    className="ds-c-vertical-nav__label"
                                    href="/logout"
                                    navigate={[Function]}
                                    onClick={[Function]}
                                  >
                                    <a
                                      className="ds-c-vertical-nav__label"
                                      href="/logout"
                                      onClick={[Function]}
                                    >
                                      Log out
                                    </a>
                                  </LinkAnchor>
                                </Link>
                              </NavLink>
                            </VerticalNavItemLabel>
                          </li>
                        </VerticalNavItem>
                      </ul>
                    </nav>
                  </VerticalNav>
                </li>
              </VerticalNavItem>
            </ul>
          </nav>
        </VerticalNav>
      </nav>
    </Nav>
  </Router>
</BrowserRouter>
`;
